src/python/pants/backend/go/util_rules/build_pkg.py

```diff
@@ -354,5 +369,54 @@ async def render_embed_config(request: RenderEmbedConfigRequest) -> RenderedEmbe
     return RenderedEmbedConfig(digest)
 
 
+# Compute a cache key for the compile action. This computation is intended to capture similar values to the
+# action ID computed by the `go` tool for its own cache.
+# For details, see https://github.com/golang/go/blob/21998413ad82655fef1f31316db31e23e0684b21/src/cmd/go/internal/work/exec.go#L216-L403
+@rule
+async def compute_compile_action_id(
+    request: GoCompileActionIdRequest, goroot: GoRoot
+) -> GoCompileActionIdResult:
+    bq = request.build_request
+
+    h = hashlib.sha256()
+
+    # All Go action IDs have the full version (as returned by `runtime.Version()` in the key.
+    # See https://github.com/golang/go/blob/master/src/cmd/go/internal/cache/hash.go#L32-L46
+    h.update(goroot.full_version.encode())
+
+    h.update("compile\n".encode())
+    if bq.minimum_go_version:
+        h.update(f"go {bq.minimum_go_version}\n".encode())
+    h.update(f"goos {goroot.goos} goarch {goroot.goarch}\n".encode())
+    h.update(f"import {bq.import_path}\n".encode())
+    # TODO: Consider what to do with this information from Go tool:
+    # fmt.Fprintf(h, "omitdebug %v standard %v local %v prefix %q\n", p.Internal.OmitDebug, p.Standard, p.Internal.Local, p.Internal.LocalPrefix)
+    # TODO: Inject cgo-related values here.
+    # TODO: Inject cover mode values here.
+    # TODO: Inject fuzz instrumentation values here.
+
+    compile_tool_id = await Get(GoSdkToolIDResult, GoSdkToolIDRequest("compile"))
+    h.update(
+        f"compile {compile_tool_id.tool_id}\n".encode()
+    )  # TODO: Add compiler flags as per `go`'s algorithm?
```

Comment:
That makes sense to me. What's the hesitation?

Reply:
No hesitation, I need to figure out _which_ options are actually incorporated. The original logic looks at two values that I still need to trace through in the `go` sources: `forcedGcflags` and  `p.Internal.Gcflags`

Reply:
And figuring that out shouldn't hold up this PR.

Reply:
Sg. Agreed, given that this PR makes things much better than before + lays a good foundation for more improvement. Maybe open a GitHub issue tracking the followup improvements so that we don't forget?

Reply:
https://github.com/pantsbuild/pants/issues/14664 and modified comments.